File: app/src/main/java/comm/example/rane22sau/integrated_level_1/Move_dist.java
package comm.example.rane22sau.integrated_level_1;

import android.Manifest;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import android.view.SurfaceView;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.Switch;
import android.widget.TextView;

import org.opencv.android.BaseLoaderCallback;
import org.opencv.android.CameraBridgeViewBase;
import org.opencv.android.JavaCameraView;
import org.opencv.android.LoaderCallbackInterface;
import org.opencv.android.OpenCVLoader;
import org.opencv.core.Core;
import org.opencv.core.CvType;
import org.opencv.core.Mat;
import org.opencv.core.MatOfPoint;
import org.opencv.core.Scalar;
import org.opencv.imgproc.Imgproc;

import java.util.ArrayList;
import java.util.List;

import static android.content.pm.PackageManager.PERMISSION_DENIED;

public class Move_dist extends MainActivity implements CameraBridgeViewBase.CvCameraViewListener2 {

    public static int a = 0, b = 0, c = 0, d = 0, e = 0, f = 0;
    int send = 0;
    Switch connectbt;
    double finalDistance;
    double t = 0, area = 3.14 * 4;
    SeekBar h_min, h_max, v_min, v_max, s_min, s_max;
    global_variables gv;
    Button red, blue, green;
    TextView text;
    private JavaCameraView javaCameraView;
    private Mat B, F;
    //te=findViewById(R.id.dist);
    private BaseLoaderCallback baseLoaderCallback = new BaseLoaderCallback(this) {
        @Override
        public void onManagerConnected(int status) {
            switch (status) {
                case BaseLoaderCallback.SUCCESS:
                    javaCameraView.enableView();
                    break;
                default:
                    super.onManagerConnected(status);
                    break;
            }
        }
    };

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        Intent b = new Intent(getApplicationContext(), demo.class);
        startActivity(b);
    }

    @Override
    protected void onResume() {
        super.onResume();

    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Intent BTconnect = new Intent(getApplicationContext(), MainActivity.class);
        if (global_variables.BT == 0) {
            startActivity(BTconnect);

        }


        setContentView(R.layout.activity_move_dist);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);


        red = (Button) findViewById(R.id.Red);
        blue = (Button) findViewById(R.id.Blue);
        green = (Button) findViewById(R.id.Green);
        text = findViewById(R.id.dist);
        text.setText(String.valueOf(t));

        red.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                /*h_min.setProgress(0); REAL_RED Color
                h_max.setProgress(10);
                s_min.setProgress(150);
                s_max.setProgress(255);
                v_min.setProgress(150);
                v_max.setProgress(255);*/
                h_min.setProgress(161);
                h_max.setProgress(188);
                s_min.setProgress(121);
                s_max.setProgress(255);
                v_min.setProgress(194);
                v_max.setProgress(255);


            }
        });
        blue.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                h_min.setProgress(240);
                h_max.setProgress(255);
                s_min.setProgress(100);
                s_max.setProgress(255);
                v_min.setProgress(100);
                v_max.setProgress(255);
            }
        });
        green.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                h_min.setProgress(120);
                h_max.setProgress(150);
                s_min.setProgress(100);
                s_max.setProgress(255);
                v_min.setProgress(50);
                v_max.setProgress(155);


            }
        });


        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PERMISSION_DENIED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA}, 101);
        }

        javaCameraView = (JavaCameraView) findViewById(R.id.javaCameraView);
        javaCameraView.setVisibility(SurfaceView.VISIBLE);
        javaCameraView.setCvCameraViewListener(this);

        h_min = (SeekBar) findViewById(R.id.h_min);
        h_max = (SeekBar) findViewById(R.id.h_max);
        v_min = (SeekBar) findViewById(R.id.v_min);
        v_max = (SeekBar) findViewById(R.id.v_max);
        s_min = (SeekBar) findViewById(R.id.s_min);
        s_max = (SeekBar) findViewById(R.id.s_max);

        h_max.setMax(255);
        h_min.setMax(255);
        v_max.setMax(255);
        v_min.setMax(255);
        s_max.setMax(255);
        s_min.setMax(255);
        //Core.inRange(B,new Scalar(gv.a,gv.c,gv.b), new Scalar(gv.d,gv.e,gv.f),B);
        h_min.setProgress(global_variables.getA());
        h_max.setProgress(global_variables.getD());
        s_min.setProgress(global_variables.getC());
        s_max.setProgress(global_variables.getE());
        v_min.setProgress(global_variables.getB());
        v_max.setProgress(global_variables.getF());

    }

    @Override
    public void onCameraViewStarted(int width, int height) {
        B = new Mat(width, height, CvType.CV_8UC4);
        F = new Mat(width, height, CvType.CV_8UC1);
        Log.d("Before", String.valueOf(B));
    }

    @Override
    public void onCameraViewStopped() {
        B.release();
        F.release();
    }

    @Override
    public Mat onCameraFrame(CameraBridgeViewBase.CvCameraViewFrame inputFrame) {

        double distance = 0;
        global_variables.a = h_min.getProgress();
        global_variables.b = v_min.getProgress();
        global_variables.c = s_min.getProgress();
        global_variables.d = h_max.getProgress();
        global_variables.e = v_max.getProgress();
        global_variables.f = s_max.getProgress();
        Log.d("TAG", "" + global_variables.a + " " + global_variables.b + " " + global_variables.c + " " + global_variables.d + " " + global_variables.e + " " + global_variables.f);
        F = inputFrame.rgba();
        Imgproc.cvtColor(F, B, Imgproc.COLOR_RGB2HSV);
        Scalar lowerThreshold = new Scalar(0, 0, 0); // Blue color – lower hsv values
        Scalar upperThreshold = new Scalar(0, 0, 255); // Blue color – higher hsv values
        //Core.inRange (B, lowerThreshold , upperThreshold, B );
        Core.inRange(B, new Scalar(global_variables.a, global_variables.b, global_variables.c), new Scalar(global_variables.d, global_variables.e, global_variables.f), B);
        Imgproc.dilate(B, B, new Mat());
        Imgproc.dilate(F, F, new Mat());
        Mat hierarchy = new Mat();
        List<MatOfPoint> contours = new ArrayList<>();
        Imgproc.findContours(B, contours, hierarchy, Imgproc.RETR_LIST, Imgproc.CHAIN_APPROX_SIMPLE);


        Mat current_contour, largest_contour = null;
        double largest_area = 0;
        int largest_contour_index = 0;

        for (int contourIdx = 0; contourIdx < contours.size(); contourIdx++) {
            current_contour = contours.get(contourIdx);
            double contourArea = Imgproc.contourArea(current_contour);
            //Imgproc.drawContours ( F, contours, contourIdx,new Scalar(255, 255, 0, 255),3);
            if (contourArea > largest_area) {
                largest_area = contourArea;
                largest_contour_index = contourIdx;
                largest_contour = current_contour;

            }
        }
        Imgproc.drawContours(F, contours, largest_contour_index, new Scalar(255, 255, 0, 255), 3);

        t = largest_area;
        //distance=422.3657139*Math.pow((3.14*4)/largest_area,0.5);
        distance = 1301.73836155 * Math.pow((area) / t, 0.5);
        final double temp = largest_area;
        //distance=26*Math.pow(330/largest_area,0.5);
        //Log.d("TAG","pixel area "+temp);
        //Log.d("TAG", "Area  "+largest_contour);

        Log.d("TAG", "DISTANCE   " + distance);
        t = distance;
        //Toast.makeText(getApplicationContext(), (int) distance,Toast.LENGTH_SHORT).show();
        finalDistance = (int) distance;
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                text.setText(String.valueOf(finalDistance));
            }
        });
        if (finalDistance <= 30 && finalDistance >= 25) {
            Intent i = new Intent(getApplicationContext(), Main3Activity.class);
            Bundle v = new Bundle();
            /*v.putInt("a",a);
            v.putInt("b",b);
            v.putInt("c",c);
            v.putInt("d",d);
            v.putInt("e",e);
            v.putInt("f",f);
            i.putExtra("masking_values",v);
*/
            startActivity(i);
            finish();
        }


        return F;
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (javaCameraView != null)
            javaCameraView.disableView();
    }

    @Override
    protected void onPostResume() {
        super.onPostResume();
        if (OpenCVLoader.initDebug()) {
            baseLoaderCallback.onManagerConnected(LoaderCallbackInterface.SUCCESS);
        } else {
            OpenCVLoader.initAsync(OpenCVLoader.OPENCV_VERSION_3_4_0, this, baseLoaderCallback);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (javaCameraView != null)
            javaCameraView.disableView();
    }
}

File: app/src/main/java/com/vitpunerobotics/netra/activities/Move_dist.kt
package com.vitpunerobotics.netra.activities

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.util.Log
import android.view.SurfaceView
import android.view.View
import android.view.WindowManager
import android.widget.Button
import android.widget.SeekBar
import android.widget.TextView
import com.vitpunerobotics.netra.R
import com.vitpunerobotics.netra.global_variables
import com.vitpunerobotics.netra.global_variables.bT
import org.opencv.android.BaseLoaderCallback
import org.opencv.android.CameraBridgeViewBase.CvCameraViewFrame
import org.opencv.android.CameraBridgeViewBase.CvCameraViewListener2
import org.opencv.android.JavaCameraView
import org.opencv.android.LoaderCallbackInterface
import org.opencv.android.OpenCVLoader
import org.opencv.core.*
import org.opencv.imgproc.Imgproc
import java.util.*

class Move_dist : MainActivity(), CvCameraViewListener2 {


    var finalDistance = 0.0
    var t = 0.0
    var area = 3.14 * 4
    var h_min: SeekBar? = null
    var h_max: SeekBar? = null
    var v_min: SeekBar? = null
    var v_max: SeekBar? = null
    var s_min: SeekBar? = null
    var s_max: SeekBar? = null

    var text: TextView? = null
    private lateinit var javaCameraView: JavaCameraView
    private var B: Mat? = null
    private var F: Mat? = null


    private val baseLoaderCallback: BaseLoaderCallback = object : BaseLoaderCallback(this) {
        override fun onManagerConnected(status: Int) {
            when (status) {
                SUCCESS -> javaCameraView.enableView()
                else -> super.onManagerConnected(status)
            }
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        val b = Intent(applicationContext, demo::class.java)
        startActivity(b)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val BTconnect = Intent(applicationContext, MainActivity::class.java)
        if (bT == 0) {
            startActivity(BTconnect)
        }
        setContentView(R.layout.activity_move_dist)
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)
        var red = findViewById<Button>(R.id.Red)
        var blue = findViewById<Button>(R.id.Blue)
        var green = findViewById<Button>(R.id.Green)
        text = findViewById(R.id.dist)
        text!!.text = t.toString()
        red.setOnClickListener(View.OnClickListener {
            h_min!!.progress = 161
            h_max!!.progress = 188
            s_min!!.progress = 121
            s_max!!.progress = 255
            v_min!!.progress = 194
            v_max!!.progress = 255
        })
        blue.setOnClickListener(View.OnClickListener {
            h_min!!.progress = 240
            h_max!!.progress = 255
            s_min!!.progress = 100
            s_max!!.progress = 255
            v_min!!.progress = 100
            v_max!!.progress = 255
        })
        green.setOnClickListener(View.OnClickListener {
            h_min!!.progress = 120
            h_max!!.progress = 150
            s_min!!.progress = 100
            s_max!!.progress = 255
            v_min!!.progress = 50
            v_max!!.progress = 155
        })
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_DENIED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.CAMERA), 101)
        }
        javaCameraView = findViewById(R.id.javaCameraView)
        javaCameraView.visibility = SurfaceView.VISIBLE
        javaCameraView.setCvCameraViewListener(this)
        h_min = findViewById(R.id.h_min)
        h_max = findViewById(R.id.h_max)
        v_min = findViewById(R.id.v_min)
        v_max = findViewById(R.id.v_max)
        s_min = findViewById(R.id.s_min)
        s_max = findViewById(R.id.s_max)
        h_max!!.max = 255
        h_min!!.max = 255
        v_max!!.max = 255
        v_min!!.max = 255
        s_max!!.max = 255
        s_min!!.max = 255
        //Core.inRange(B,new Scalar(gv.a,gv.c,gv.b), new Scalar(gv.d,gv.e,gv.f),B);
        h_min!!.progress = global_variables.a
        h_max!!.progress = global_variables.d
        s_min!!.progress = global_variables.c
        s_max!!.progress = global_variables.e
        v_min!!.progress = global_variables.b
        v_max!!.progress = global_variables.f
    }

    override fun onCameraViewStarted(width: Int, height: Int) {
        B = Mat(width, height, CvType.CV_8UC4)
        F = Mat(width, height, CvType.CV_8UC1)
        Log.d("Before", B.toString())
    }

    override fun onCameraViewStopped() {
        B!!.release()
        F!!.release()
    }

    override fun onCameraFrame(inputFrame: CvCameraViewFrame): Mat {
        var distance = 0.0
        global_variables.a = h_min!!.progress
        global_variables.b = v_min!!.progress
        global_variables.c = s_min!!.progress
        global_variables.d = h_max!!.progress
        global_variables.e = v_max!!.progress
        global_variables.f = s_max!!.progress
        Log.d("TAG", "" + global_variables.a + " " + global_variables.b + " " + global_variables.c + " " + global_variables.d + " " + global_variables.e + " " + global_variables.f)
        F = inputFrame.rgba()
        Imgproc.cvtColor(F, B, Imgproc.COLOR_RGB2HSV)
        val lowerThreshold = Scalar(0.0, 0.0, 0.0) // Blue color – lower hsv values
        val upperThreshold = Scalar(0.0, 0.0, 255.0) // Blue color – higher hsv values
        Core.inRange(B, Scalar(global_variables.a.toDouble(), global_variables.b.toDouble(), global_variables.c.toDouble()), Scalar(global_variables.d.toDouble(), global_variables.e.toDouble(), global_variables.f.toDouble()), B)
        Imgproc.dilate(B, B, Mat())
        Imgproc.dilate(F, F, Mat())
        val hierarchy = Mat()
        val contours: List<MatOfPoint> = ArrayList()
        Imgproc.findContours(B, contours, hierarchy, Imgproc.RETR_LIST, Imgproc.CHAIN_APPROX_SIMPLE)
        var current_contour: Mat
        var largest_contour: Mat? = null
        var largest_area = 0.0
        var largest_contour_index = 0
        for (contourIdx in contours.indices) {
            current_contour = contours[contourIdx]
            val contourArea = Imgproc.contourArea(current_contour)

            if (contourArea > largest_area) {
                largest_area = contourArea
                largest_contour_index = contourIdx
                largest_contour = current_contour
            }
        }
        Imgproc.drawContours(F, contours, largest_contour_index, Scalar(255.0, 255.0, 0.0, 255.0), 3)
        t = largest_area

        distance = 1301.73836155 * Math.pow(area / t, 0.5)
        val temp = largest_area

        Log.d("TAG", "DISTANCE   $distance")
        t = distance
        //Toast.makeText(getApplicationContext(), (int) distance,Toast.LENGTH_SHORT).show();
        finalDistance = distance
        runOnUiThread { text!!.text = finalDistance.toString() }
        if (finalDistance <= 30 && finalDistance >= 25) {
            val i = Intent(applicationContext, Main3Activity::class.java)
            val v = Bundle()

            startActivity(i)
            finish()
        }
        return F!!
    }

    override fun onPause() {
        super.onPause()
        javaCameraView.disableView()
    }

    override fun onPostResume() {
        super.onPostResume()
        if (OpenCVLoader.initDebug()) {
            baseLoaderCallback.onManagerConnected(LoaderCallbackInterface.SUCCESS)
        } else {
            OpenCVLoader.initAsync(OpenCVLoader.OPENCV_VERSION_3_4_0, this, baseLoaderCallback)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        javaCameraView.disableView()
    }



}


